libs/wingc/src/type_check.rs

```diff
@@ -1865,399 +1883,416 @@ impl<'a> TypeChecker<'a> {
-				}
-				Literal::Number(_) => (self.types.number(), Phase::Independent),
-				Literal::Boolean(_) => (self.types.bool(), Phase::Independent),
-			},
-			ExprKind::Binary { op, left, right } => {
-				let (ltype, ltype_phase) = self.type_check_exp(left, env);
-				let (rtype, _) = self.type_check_exp(right, env);
-
-				match op {
-					BinaryOperator::LogicalAnd | BinaryOperator::LogicalOr => {
-						self.validate_type(ltype, self.types.bool(), left);
-						self.validate_type(rtype, self.types.bool(), right);
-						(self.types.bool(), Phase::Independent)
+	fn type_check_exp_ext(
+		&mut self,
+		exp: &Expr,
+		env: &mut SymbolEnv,
+		expr_visit_info: ExprVisitInfo,
+	) -> (TypeRef, Phase) {
+		CompilationContext::set(CompilationPhase::TypeChecking, &exp.span);
+
+		self.curr_expr_info.push(expr_visit_info);
+
+		//let (mut t, phase) = self.type_check_exp_helper(&exp, env);
+		let (mut t, phase) = |exp: &Expr, env: &mut SymbolEnv| -> (TypeRef, Phase) {
+			match &exp.kind {
+				ExprKind::Literal(lit) => match lit {
+					Literal::String(_) => (self.types.string(), Phase::Independent),
+					Literal::Nil => (self.types.nil(), Phase::Independent),
+					Literal::InterpolatedString(s) => {
+						let mut phase = Phase::Independent;
+						s.parts.iter().for_each(|part| {
+							if let InterpolatedStringPart::Expr(interpolated_expr) = part {
+								let (exp_type, p) = self.type_check_exp(interpolated_expr, env);
+								phase = Self::combine_phases(phase, p);
+								self.validate_type_in(exp_type, &self.types.stringables(), interpolated_expr);
+							}
+						});
+						(self.types.string(), phase)
 					}
-					BinaryOperator::AddOrConcat => {
-						if ltype.is_subtype_of(&self.types.number()) && rtype.is_subtype_of(&self.types.number()) {
-							(self.types.number(), Phase::Independent)
-						} else if ltype.is_subtype_of(&self.types.string()) && rtype.is_subtype_of(&self.types.string()) {
-							(self.types.string(), Phase::Independent)
-						} else {
-							// If any of the types are unresolved (error) then don't report this assuming the error has already been reported
-							if !ltype.is_unresolved() && !rtype.is_unresolved() {
-								self.spanned_error(
-									exp,
-									format!(
-										"Binary operator '+' cannot be applied to operands of type '{}' and '{}'; only ({}, {}) and ({}, {}) are supported",
-										ltype,
-										rtype,
-										self.types.number(),
-										self.types.number(),
-										self.types.string(),
-										self.types.string(),
-									),
-								);
+					Literal::Number(_) => (self.types.number(), Phase::Independent),
+					Literal::Boolean(_) => (self.types.bool(), Phase::Independent),
+				},
+				ExprKind::Binary { op, left, right } => {
+					let (ltype, ltype_phase) = self.type_check_exp(left, env);
+					let (rtype, rtype_phase) = self.type_check_exp(right, env);
+
+					// Resolve the phase
+					let phase = Self::combine_phases(ltype_phase, rtype_phase);
```

Comment:
Do we need to combine phases for array literal / collection literal expressions?

Reply:
Good point. The code for these literals currently adopts the phase of the env (and doesn't default to `Phase::Independent`). This means that this will generate and error:
```wing
// preflight
let bucket = new cloud.Bucket();
//inflight
assert([bucket].at(0).list().length == 0);
//     ^ this is considered an inflight expr, but we can fix this to be a preflight expr and then it can be lifted
```
I'll open a tracking issue for this, because it doesn't seem critical.
The critical issue is if we can produce an inflight expr that the compile won't detect as inflight and then it might try to lift something that it can't. That's why I created the `combile_pahses` logic. To deal with this case:
```
// inflight
preflight_array.at(0 + inflight_var);
//                 ^^^^^^^^^^^^^^^^ this was considered before as phase independent and then we didn't know that `at` can't be lifted here.
```

Reply:
See #5359